### DISCUSSION TITLE
Allow developers to specify additional paths for output file tracing

### DISCUSSION THREAD

**Discussion by majelbstoat** [vercel/next.js#35792] [Ideas]
### Describe the feature you'd like to request

Because Next doesn't offer a way to run tasks just before server startup, developers need to resort to `NODE_OPTIONS="--require preload.js"` to  support certain tasks like open telemetry initialisation.

Unfortunately, this `preload.js` file and it dependencies will not be traced as part of outputFileTracing. We can write a standalone script that mimics `trace-next-server`, but it is largely duplicative.

### Describe the solution you'd like

`trace-next-server` makes a call to `nodeFileTrace` with an array comprising a single path to `nextServerPath`:

https://github.com/vercel/next.js/blob/ff474b01db0ee2ed176b046fc2d0c891bc4efa88/scripts/trace-next-server.js#L73

I would like to propose a new configuration parameter `outputFileTracingExtraPaths`, an array of paths to additional paths that will also be traced.

This array would be appended to the array sent to nodeFileTrace, ensuring that the files and their dependencies make it into next-server-js.nft.json

The paths would also be included in the standalone bundle.

### Describe alternatives you've considered

* Writing our own complementary version of `trace-next-server` (our current approach): lots of duplication, and keeping it up to date is brittle.
* `unstable_includePaths: ['preload.js']` in the export of `pages/index.tsx`: this includes the preload file itself in the standalone output, but none of its `require'd` dependencies. it's also a bit strange.

**Comment by EmilioAiolfi** [comment]
How to improve telemetry approach in Nextjs without needing `NODE_OPTIONS="--require preload.js`? This with standalone mode, I know this feature is experimental, but today it is not possible to use standalone mode and have a telemetry (Datadog example).

Does the Vercel team have an opinion for this?

**Comment by Nargonath** [comment]
We use Output File Tracing as well and we had errors when setting up NextRelic. NewRelic has two packages you need to install: `newrelic` and `@newrelic/next`. The former is required in my `_document.tsx` `getInitialProps` function because I need it to inject the web agent script tag in the HTML document thus it gets included in the `standalone` folder. I tried to use the `unstable_includeFiles` page config to include the latter inside my `page/index.tsx` as:

```ts
export const config: PageConfig = {
  unstable_includeFiles: ["node_modules/@newrelic/next/**/*"]
};
```

But I get an error `EISDIR` as `next` tries to use `copyFile` on a directory:
```
> Build error occurred
[Error: EISDIR: illegal operation on a directory, copyfile '/home/user/frontend/node_modules/@newrelic/next/lib' -> '/home/user/frontend/.next/standalone/node_modules/@newrelic/next/lib'] {
  errno: -21,
  code: 'EISDIR',
  syscall: 'copyfile',
  path: '/home/user/frontend/node_modules/@newrelic/next/lib',
  dest: '/home/user/frontend/.next/standalone/node_modules/@newrelic/next/lib'
}
```

My workaround for now is to actually load also `@newrelic/next` inside my `getInitialProps` in `_document.tsx` as:
```ts
require("@newrelic/next")
```
This way it gets included by `next`.

**Reply by henrych4** [reply]
Got same error, it works after changed to /\*.\* instead of /*

ie.
```javascript
export const config: PageConfig = {
  unstable_includeFiles: ["node_modules/@newrelic/next/**/*.*"]
};
```
